provider/provider.go

```diff
@@ -113,3 +113,60 @@ func pinSet(ctx context.Context, pinning pin.Pinner, fetchConfig fetcher.Factory
 
 	return set, nil
 }
+
+func NewPrioritizedProvider(priorityCids KeyChanFunc, otherCids KeyChanFunc) KeyChanFunc {
+	return func(ctx context.Context) (<-chan cid.Cid, error) {
+		outCh := make(chan cid.Cid)
+
+		go func() {
+			defer close(outCh)
+			visited := map[string]struct{}{}
+
+			handleStream := func(stream KeyChanFunc, markVisited bool) error {
+				ch, err := stream(ctx)
+				if err != nil {
+					return err
+				}
+
+				for {
+					select {
+					case <-ctx.Done():
+						return nil
+					case c, ok := <-ch:
+						if !ok {
+							return nil
+						}
+
+						// Likely faster than c.String()
+						str := string(c.Bytes())
```

Comment:
c.KeyString()?

Reply:
Thinking a bit it might be more sensible to use the multihash bytes here rather than the CID bytes given that the major routing systems in use (Amino DHT, IPNI) are multihash based rather than CID based and so advertising 2 CIDs with the same multihash isn't really useful anyhow.

If people feel this abstraction should work on CIDs it's probably not too wasteful for most users, but I doubt changing will hurt anyone given routing systems that want more than just multihashes probably want more information than just CIDs (unless they're only differentiating full-DAG vs block)